### PR TITLE
Kectl Cloud init and reset code

### LINKED ISSUE #360: Can we add a node by a single command?
**What would you like to be added**:
Can we add a node by a single command like “k3s agent -s ${SERVER_URL} -t ${NODE_TOKEN} ” in k3s, or “docker swarm join-token ${NODE_TOKEN}” in docker swarm?
**Why is this needed**:
The operation is simple and common.

### PR DESCRIPTION
**Pull request by srivatsav123**
**What type of PR is this?**
Implementation of cloud init and reset

 /kind feature

**What this PR does / why we need it**:
Implements cloud Init and Reset
**Which issue(s) this PR fixes**:
Fixes #342 

*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

**Special notes for your reviewer**:
Special notes for your reviewer:

    1.This PR is raised for cloud node commands like, "kectl cloud init" and "kectl cloud reset" on ubuntu OS.
    2.Code comments are yet to be added every where.
    3.Error handling and some function(check apiserver health ) must be implemented and is in progress
    4.Cloud commands like "kectl cloud init" and "kectl cloud reset" will be added to this PR soon.
    5.There will be lot of debug fmt.Println's, please ignore them as of now. It will be removed soon.
    6. Lints will fail, so please ignore for now.
    7.E2E will be soon tested
    8. These PR must be merged ,after  edge join and edge reset PR is merged
    9. Certgen.sh script must be placed in release package(kubeedge-v0.2.1-linux-amd64.tar.gz) in /etc/kubeedge/kubeedge/tools folder otherwise installation fails for first time saying certificates not generated

### CODE REVIEW

**Comment by srivatsav123** [comment]
@samy2019 -Request you to Please remove cloud/init.go and cloud/reset.go (skeleton code from your PR)..

File: kectl/app/cmd/cloud/init.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+var (
+	cloudInitLongDescription = `
+kectl init command bootstraps KubeEdge's cloud component.
+It checks if the pre-requisites are installed already,
+If not installed, this command will help in download,
+install and execute on the host.
+`
+	cloudInitExample = `
+kectl cloud init  --docker-version= --kubernetes-version=<version> --kubeedge-version<version> --docker-version=<version>
```
Comment (rohitsardesai83) [inline kectl/app/cmd/cloud/init.go]: docker-version is appearing twice. please correct.

Comment (rohitsardesai83) [reply kectl/app/cmd/cloud/init.go]: --kubeedge-version<version>  -> --kubeedge-version=<version>

File: kectl/app/cmd/cloud/init.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+type FlagData struct {
+	Val    interface{}
+	DefVal interface{}
+}
+
+// NewCloudInit represents the kubeedge cloud init command
+func NewCloudInit(out io.Writer, init *options.InitOptions) *cobra.Command {
+	if init == nil {
+		init = newInitOptions()
+		fmt.Println("init is %v", init)
```
Comment (rohitsardesai83) [inline kectl/app/cmd/cloud/init.go]: please remove all fmt.println code later.

File: kectl/app/cmd/util/cloudinstaller.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+//kubeadm reset will be called
+func (u *KubeCloudInstTool) ResetKubernetes() error {
+	binExec := fmt.Sprintf("echo 'y' | kubeadm reset && sudo rm -rf ~/.kube")
+	cmd := &Command{Cmd: exec.Command("sh", "-c", binExec)}
+	err := cmd.ExecuteCmdShowOutput()
+	errout := cmd.GetStdErr()
+	if err != nil || errout != "" {
+		return fmt.Errorf("kubernetes installation failed %s", errout)
```
Comment (rohitsardesai83) [inline kectl/app/cmd/util/cloudinstaller.go]: installation - > reset

File: kectl/app/cmd/util/cloudinstaller.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+		errout = cmd.GetStdErr()
+		if err != nil || errout != "" {
+			fmt.Errorf("kubernetes Installation failed:%s", stdout)
+		}
+	}
+	return nil
+}
+
+//Kubernetes Manifests file will be updated by necessary parameters
+func (cu *KubeCloudInstTool) updatemanifests() error {
```
Comment (rohitsardesai83) [inline kectl/app/cmd/util/cloudinstaller.go]: follow camelCasing for all method names.

File: kectl/app/cmd/cloud/reset.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+			TearDownEdgeNode()
+		},
+		Example: cloudResetExample,
+		Args:    cobra.NoArgs,
+	}
+
+	return cmd
+}
+
+//TearDoen nodes will do kubeadm reset and kill edgecontroller
```
Comment (rohitsardesai83) [inline kectl/app/cmd/cloud/reset.go]: TearDoen -> TearDown

File: kectl/app/cmd/cloud/reset.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+func TearDownEdgeNode() {
+	cloud := &util.KubeCloudInstTool{Common: util.Common{}}
+	cloud.ResetKubernetes()
+	cloud.KillEdgeController()
```
Comment (rohitsardesai83) [inline kectl/app/cmd/cloud/reset.go]: I think first edge controller should be killed and then kubeadm reset

File: kectl/app/cmd/common.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+//Constants used by installers
+const (
+	KubeCloudCertificatePath   = "/etc/kubeedge/kubeedge/certificates/"
+	KubeControllerConfig       = "/etc/kubeedge/kubeedge/cloud/conf/controller.yaml"
+	KubeCloudBinaryName        = "edgecontroller"
+	KubeCloudApiserverYamlPath = "/etc/kubernetes/manifests/kube-apiserver.yaml"
+	KubeCloudReplaceIndex      = 25
+	KubeCloudReplaceString     = "    - --insecure-bind-address=0.0.0.0\n"
```
Comment (rohitsardesai83) [inline kectl/app/cmd/common.go]: This could be renamed to match the value which is insecure bind address.

File: kectl/app/cmd/common.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+type ToolsInstallerCloud interface {
```
Comment (rohitsardesai83) [inline kectl/app/cmd/common.go]: please add sufficient documentation for all the interfaces and apis which we expose through the installer.

File: kectl/app/cmd/util/cloudinstaller.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+	cu.SetOSInterface(GetOSInterface())
+	cu.SetKubeEdgeVersion(cu.ToolVersion)
+
+	err := cu.InstallKubeEdge()
+	if err != nil {
+
+		return err
+	}
+	fmt.Println("Installation of kubeedge package is sucessfull")
+	err = cu.generatecertificates()
```
Comment (rohitsardesai83) [inline kectl/app/cmd/util/cloudinstaller.go]: camelCasing for function name

File: kectl/app/cmd/util/cloudinstaller.go [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+	if err != nil {
+		fmt.Println(" in err")
+		return err
+	}
+	fmt.Println("Certificates got genertaed and its kept in /etc/kubeedge/kubeedge certificates folder")
```
Comment (rohitsardesai83) [inline kectl/app/cmd/util/cloudinstaller.go]: genertaed -> generated

Comment (rohitsardesai83) [reply kectl/app/cmd/util/cloudinstaller.go]: This log could be shortened : Certificates got generated at : /etc/kubeedge/kubeedge

File: kectl/app/cmd/util/cloudinstaller.go (hunk 9 above)
Comment (rohitsardesai83) [inline kectl/app/cmd/util/cloudinstaller.go]: Can we come up with a phase for every step and the code related to that goes in that phase go file? 

see https://github.com/kubernetes/kubernetes/tree/master/cmd/kubeadm/app/phases for reference.

File: kectl/app/cmd/util/cloudinstaller.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+	err = cu.tarcertificates()
+	if err != nil {
+		return err
+	}
+	fmt.Println("Certificates got Tared and kept /etc/kubeedge/kubeedge/certificates folder Please copy the certificates to the respective edge node")
```
Comment (rohitsardesai83) [inline kectl/app/cmd/util/cloudinstaller.go]: Please use proper / correct vocabulary along with proper periods , commas etc.

File: kectl/app/cmd/util/cloudinstaller.go (hunk 9 above)
Comment (srivatsav123) [reply kectl/app/cmd/util/cloudinstaller.go]: https://github.com/srivatsav123/kubeedge/tree/kubeedgeinstaller/kubeedgeinst/cmd/phases/workflow
Actually if u see my skeleton code....I had added phases(I had a  thought on it)...Sameer had refractored the frame work, so i am going with the current implementation ...Will discuss with sameer about it..

File: kectl/app/cmd/cloud/reset.go (hunk 6 above)
Comment (srivatsav123) [reply kectl/app/cmd/cloud/reset.go]: ok

**Comment by srivatsav123** [comment]
cloud init
init1
init2
Screenshot from 2019-04-30 21-37-27


cloud reset
reset


E2E testing results



nodepod-status

File: kectl/app/cmd/cloud/init.go [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+	addJoinOtherFlags(cmd, init)
+	return cmd
+}
+
+//newInitOptions will initialise new instance of options everytime
+func newInitOptions() *options.InitOptions {
+	var opts *options.InitOptions
+	opts = &options.InitOptions{}
+	opts.DockerVersion = options.DefaultDockerVersion
+	opts.KubeedgeVersion = options.DefaultKubeEdgeVersion
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: Kubeedge -> KubeEdge

File: kectl/app/cmd/cloud/init.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+	opts.Kubernetesversion = options.DefaultK8SVersion
+
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: please remove blank line

File: kectl/app/cmd/cloud/init.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+	return opts
+}
+
+func addJoinOtherFlags(cmd *cobra.Command, initOpts *options.InitOptions) {
+
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: please remove blank line

File: kectl/app/cmd/cloud/init.go [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+	cmd.Flags().StringVar(&initOpts.KubeedgeVersion, options.KubeedgeVersion, initOpts.KubeedgeVersion,
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: Kubeedge -> KubeEdge
Applicable to all similar places

File: kectl/app/cmd/cloud/init.go [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+	cmd.Flags().StringVar(&initOpts.Kubernetesversion, options.Kubernetesversion, initOpts.Kubernetesversion,
+		"Use this key to download and use the required Kubernetes version")
+}
+
+func Add2ToolsList(toolList map[string]util.ToolsInstallerCloud, flagData map[string]FlagData, initOptions *options.InitOptions) {
+	var kubeVer, dockerVer, k8sVer string
+
+	flgData, ok := flagData[options.KubeedgeVersion]
+	if ok {
+		fmt.Println(options.KubeedgeVersion, "VAL:", flgData.Val.(string), "DEFVAL:", flgData.DefVal.(string))
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: Please remove all fmt.Print statements and add logs

File: kectl/app/cmd/cloud/init.go [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+	flgData, ok = flagData[options.Kubernetesversion]
+	if ok {
+		fmt.Println(options.Kubernetesversion, "VAL:", flgData.Val.(string), "DEFVAL:", flgData.DefVal.(string))
+		k8sVer = CheckIfAvailable(flgData.Val.(string), flgData.DefVal.(string))
+	} else {
+		k8sVer = initOptions.Kubernetesversion
+	}
+	fmt.Println(options.Kubernetesversion, "VAL:", k8sVer, "DEFVAL:", flgData.DefVal.(string))
+	toolList["Kubernetes"] = &util.K8SInstTool{Common: util.Common{ToolVersion: k8sVer}, IsEdgeNode: false, DefaultToolVer: flgData.DefVal.(string)}
+
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: please remove blank line

File: kectl/app/cmd/cloud/init.go [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+func AddToolVals(f *pflag.Flag, flagData map[string]FlagData) {
+	flagData[f.Name] = FlagData{Val: f.Value.String(), DefVal: f.DefValue}
+}
+
+//Install all the required tools
+func Execute(toolList map[string]util.ToolsInstallerCloud) {
+
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: please remove blank line at the starting and ending of functions.
Applicable to all similar places

File: kectl/app/cmd/cloud/reset.go [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+		Long:  cloudResetLongDescription,
+		Run: func(cmd *cobra.Command, args []string) {
+			// TODO: Work your own magic here
+			fmt.Println("cloud reset called")
+			TearDownCloud()
+		},
+		Example: cloudResetExample,
+		Args:    cobra.NoArgs,
+	}
+
```
Comment (sids-b) [inline kectl/app/cmd/cloud/reset.go]: please remove blank line

File: kectl/app/cmd/util/cloudinstaller.go [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+		return err
+	}
+	fmt.Println("Certificates got generated at : /etc/kubeedge/kubeedge")
+	err = cu.tarCertificates()
+	if err != nil {
+		return err
+	}
+	fmt.Println("Certificates got Tared at : /etc/kubeedge/kubeedge/certificates folder, Please copy the certificates to the respective edge node")
+
+	err = cu.startKubernetescluster()
```
Comment (sids-b) [inline kectl/app/cmd/util/cloudinstaller.go]: rename function startKubernetescluster to startKubernetesCluster.
Use camel case please

File: kectl/app/cmd/util/cloudinstaller.go [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+	Common
+}
+
+func (cu *KubeCloudInstTool) InstallTools() error {
+	cu.SetOSInterface(GetOSInterface())
+	cu.SetKubeEdgeVersion(cu.ToolVersion)
+
+	err := cu.InstallKubeEdge()
+	if err != nil {
+
```
Comment (sids-b) [inline kectl/app/cmd/util/cloudinstaller.go]: please remove blank line

File: kectl/app/cmd/util/cloudinstaller.go [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+	if err != nil || errout != "" {
+		return fmt.Errorf("%s", "certificates not tared")
+		fmt.Println("in error")
+	}
+	fmt.Println(stdout)
+	return nil
+}
+
+//controller yaml ca and cert path will be replaces
+func (cu *KubeCloudInstTool) updateControlleryaml() error {
```
Comment (sids-b) [inline kectl/app/cmd/util/cloudinstaller.go]: camel case please.

File: kectl/app/cmd/util/cloudinstaller.go [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+func (u *KubeCloudInstTool) RunEdgeController() error {
+
+	filetoCopy := fmt.Sprintf(" cp %s/kubeedge/cloud/%s /usr/local/bin/.", KubeEdgePath, KubeCloudBinaryName)
+	cmd := &Command{Cmd: exec.Command("sh", "-c", filetoCopy)}
+	err := cmd.ExecuteCmdShowOutput()
+	errout := cmd.GetStdErr()
+	if err != nil || errout != "" {
+		fmt.Println("in error")
+		return fmt.Errorf("%s", errout)
+
```
Comment (sids-b) [inline kectl/app/cmd/util/cloudinstaller.go]: please remove blank line

File: kectl/app/cmd/util/cloudinstaller.go [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+		fmt.Println("in error")
+	}
+	fmt.Println(stdout)
+	return nil
+}
+
+//kubeadm init will be called and cluster will be started
+func (cu *KubeCloudInstTool) startKubernetescluster() error {
+	var install bool
+	cmd := &Command{Cmd: exec.Command("sh", "-c", "kubeadm version")}
```
Comment (sids-b) [inline kectl/app/cmd/util/cloudinstaller.go]: what if kubeadm nor present ?

File: kectl/app/cmd/cloud/init.go [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+	var opts *options.InitOptions
+	opts = &options.InitOptions{}
+	opts.DockerVersion = options.DefaultDockerVersion
+	opts.KubeedgeVersion = options.DefaultKubeEdgeVersion
+	opts.Kubernetesversion = options.DefaultK8SVersion
+
+	return opts
+}
+
+func addJoinOtherFlags(cmd *cobra.Command, initOpts *options.InitOptions) {
```
Comment (sids-b) [inline kectl/app/cmd/cloud/init.go]: please add comments on top of all functions

File: kectl/app/cmd/util/cloudinstaller.go [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+	}
+	fmt.Println("Updation of manifests is sucessful")
+
+	err = cu.updateControlleryaml()
+	if err != nil {
+		return err
+	}
+	fmt.Println("Updation of controller yaml is sucess")
+
+	time.Sleep(10 * time.Second)
```
Comment (sids-b) [inline kectl/app/cmd/util/cloudinstaller.go]: please remove sleep and wait till apiserver exposes port 8080. Give a timeout time that if after 10 or 15 seconds the port is still not exposed, return error

**Comment by sids-b** [comment]
Is docker and kubeadm installation out of scope of ??

**Comment by sids-b** [comment]
Also we need to see if docker and kubernetes versions conflict with each other. If there is an error during installation, the reason for error should be clearly highlighted

File: kectl/app/cmd/cloud/init.go (hunk 1 above)
Comment (srivatsav123) [reply kectl/app/cmd/cloud/init.go]: ok

File: kectl/app/cmd/cloud/init.go (hunk 2 above)
Comment (srivatsav123) [reply kectl/app/cmd/cloud/init.go]: ok

**Comment by sids-b** [comment]
As per discussion with @srivatsav123 and @samy2019 , new PR will be raised

**Pull request closed without merging**